utils/appliance/plugin.py

```diff
@@ -7,17 +7,21 @@ class AppliancePluginException(Exception):
     pass
 
 
+@attr.s(slots=True)
 class AppliancePluginDescriptor(object):
-    def __init__(self, cls, args, kwargs):
-        self.cls = cls
-        self.args = args
-        self.kwargs = kwargs
+    cls = attr.ib()
+    args = attr.ib()
+    kwargs = attr.ib()
+    cache = attr.ib(init=False, default=attr.Factory(dict), repr=False)
```

Comment:
**OPTIONAL** for a descriptor bound cache, it makes sense to use a WeakKeyDictionary from weakref